### DISCUSSION TITLE
Is there a way to search users by its full name using their firstName and lastName columns?

### DISCUSSION THREAD

**Discussion by johhansantana** [prisma/prisma#2585] [Q&A] [answered]
I was wondering if there was a way to filter, for example, users based on both their first and last name using a single arg

```
model User {
  id Int @default(autoincrement()) @id
  firstName String
  lastName String
}
```

I was thinking of having also a column with `fullName` where I programmatically add everything there as a `String` then use the `contain` filter there but I'm not sure if this is the right way?

**Comment by ryands17** [comment]
Hey @johhansantana :wave: 
Currently Prisma doesn't support DB level computed properties, there's an [open request](https://github.com/prisma/prisma/issues/678) that you could follow that would to implement this in the form of views.

As a workaround, to query by the full name, you would have to do something like this:
```ts
  const [firstName, lastName] = fullname.split(' ')
  await prisma.user.findMany({
    where: {
      firstName,
      lastName,
    },
  })
```
Let me know if that helps.

**Reply by johhansantana** [reply]
I might be able to make something for this to work.

I also saw with the new [beta release](https://github.com/prisma/prisma/releases/tag/2.0.0-beta.6) that we can use `.raw` to make sql queries as well. Maybe there's something easier there to use I might take a look.

Thanks!

**Reply by ryands17** [reply]
Yes `raw` could be used as well as a workaround for DB level computed values.

**Reply by ryands17** [reply]
@johhansantana If you're using Postgres, you can use generated columns as follows:

1. Create the table

```sql
create table "User" (
    id serial primary key,
    firstname text not null,
    lastname text,
    fullname text generated always as (firstname || ' ' || lastname) stored
)
```

2. Run `prisma introspect`

```prisma
model User {
  firstname String
  fullname  String?
  id        Int     @default(autoincrement()) @id
  lastname  String?
}
```

Now you will be able to use the `fullname` computed property directly.

**Reply by johhansantana** [reply]
**Update** see my reply [below](https://github.com/prisma/prisma/discussions/2585#discussioncomment-53105) for better answer

I ended up using something like this:

```sql
SELECT U.*
FROM public."User" U
WHERE (U.role = 'ADMIN'
    AND
       concat(concat(U."firstName", ' '), U."lastName") like '%Johhan Santana%')
   OR (U.role = 'ADMIN'
    AND
       concat(concat(U."firstName", ' '), concat(U."middleName", ' '), U."lastName") like '%Johhan Santana%');
```

Performance might suffer if there's a lot of records but for me, it works.

**Note** that I have also `middleName` there and this will only work if you search the columns in the order that you put them in the concact.

e.g. 
This works
`Johhan Santana`

this doesn't unless you add more conditions
`Santana Johhan`

I'm not a sql guru by any means lol I bet there's a way to fix this.

**Comment by johhansantana** [comment, marked as answer]
**Update**
View [this reply](https://github.com/prisma/prisma/discussions/2585#discussioncomment-59722) for further details.

**Original answer**
I've recently came back to this issue and refactor the code a bit:

```sql
SELECT U.*
FROM public."User" U
WHERE (U.role = 'ADMIN'
    AND
       lower(concat(U."firstName", U."middleName", U."lastName")) like '%santana%' AND
       lower(concat(U."firstName", U."middleName", U."lastName")) like '%johhan%'
          );
```

now it doesn't matter how many columns you want to search, you can just add them there. You can then loop through the search query, for example, `santana johhan` and create the `WHERE` clause accordingly like so:

```typescript
let searchConditions: string[] = [];
args.fullName.split(" ").forEach((separatedSearch, index) => {
  searchConditions.push(
    `lower(concat(U."firstName", U."middleName", U."lastName")) like '%${separatedSearch}%' ${
      index + 1 === args.fullName?.split(" ").length ? "" : "AND"
    }`
  );
});
``` 

and use it in `$queryRaw` like this:

```sql
SELECT U.*
FROM public."User" U
WHERE (U.role = 'ADMIN'
    AND
       ${searchConditions.join(" ")}
          );
```

**Reply by johhansantana** [reply]
Back when doing this. I didn't saw the [how to use variables documentation](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/raw-database-access#using-variables) where it explains how to safely use them.

In [this reply](https://github.com/prisma/prisma/discussions/3481#discussioncomment-59701), I have a better solution on this.

But basically, instead of using $queryRaw(` `` `) use $queryRaw`` instead

**Comment by ludralph** [comment]
Hi @johhansantana,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
